standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+
+Different parent images should be created in line with each framework's best practice.  Eg, Node.js, .Net Core, Ruby etc.
+
+Parent Dockerfiles should be hosted in GitHub and be deployed to a container registry via CI/CD.
+
+Example Node.js parent image:
+```
+ARG NODE_VERSION=10.18.0
+FROM node:$NODE_VERSION-alpine
+ARG NODE_VERSION
+ARG VERSION=1.0.0
+
+# We need a basic init process to handle signals and reap zombie processes, tini handles that
+RUN apk update && apk add --no-cache tini=0.18.0-r0
+ENTRYPOINT ["/sbin/tini", "--"]
+
+# Never run as root, use the node user from the alpine image
+USER node
+# Default workdir for all the containers that use this image
+WORKDIR /home/node
+
+ENV NODE_ENV production
+# Set global npm dependencies to be stored under the node user directory
+ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
+ENV PATH=$PATH:/home/node/.npm-global/bin
+LABEL uk.gov.defra.ffc-node-base.version=$VERSION \
+      uk.gov.defra.ffc-node-base.node-version=$NODE_VERSION
+```
+
+#### Node.js
+- extend a minimal Node base image - e.g. node:alpine
+- set a fixed major version of the base image
+- set the `NODE_ENV` environment variable to production in parent
+- set global NPM dependencies to be stored under the node user directory:
+  ```
+  ENV NPM_CONFIG_PREFIX=/home/node/.npm-global
+  ENV PATH=$PATH:/home/node/.npm-global/bin
+  ```
+- run as the `node` user and not `root`
+- ensure the working directory is owned by the node user - e.g. /home/node
+- make the exposed port configurable via a build argument and set as an environment variable:
+  ```
+  ARG PORT=3000
+  ENV PORT ${PORT}
+  ```
+
+#### .Net Core
+- a parent image should exist for both the .Net Core SDK and the .Net Core Runtime
+- extend minimal .Net Core base image
+- set a fixed major version of the base image
+- set the `ASPNETCORE_ENVIRONMENT` environment variable to production in parent
+- run as the `www-data` user and not `root`
+- ensure the working directory is owned by the www-data user - e.g. /app
+- a remote debugger should be added to the SDK file to aid development and testing
+  ```
+  RUN apt-get update \
+  && apt-get install -y --no-install-recommends unzip \
+  && curl -sSL https://aka.ms/getvsdbgsh | bash /dev/stdin -v latest -l /vsdbg
+  ```
+
+## Container scanning
+Container scanning tools collate vulnerability reports from a number of sources to determine security issues within operating system kernels, libraries, and software packages.
+
+Containers should be regularly scanned using tools such as Clair, Anchore Engine, Aqua Trivy or JFrog XRay.
+
+A comparison of the different tools is available in [Confluence](https://eaflood.atlassian.net/wiki/spaces/FPS/pages/1408172407/Container+Image+Scanning+Tools)
+
+## Multi stage builds
+Dockerfiles should implement multi stage builds to allow different container configuration and behaviour to be created for development, test and production.
```

Comment:
The down side of this approach as that you're explicitly defining _different_ images for different environments, when what we're aiming for, from a software development perspective, is to be able to release an image that we know will be consistent in whatever environment it is deployed, so reducing the "it works on my machine" problem.  If we're testing images that are different to the ones were we're developing and then deploying different images again to production, we're just creating a different form of the same problem.

It would be good to discuss this a bit further I think.

Reply:
It's not that the environment is changing, it's to make sure the environment only has what it needs, for example the final production image doesn't include all the unit tests files, the unit test image starts up running the command to run tests whereas the production image runs the command to start the application.

.Net Core is a bit more awkward as to build you need the SDK image which is a massive amount of gigabytes and extra "stuff" you wouldn't want in production so you're essentially building against a different base image.  This isn't really a limitation of what I'm proposing here though, it's just what you need to build .Net Core.  You could run the SDK in production, but yeah, don't.

Reply:
In all scenarios though you can run the Dockerfile to completion without targeting a build stage and end up with exactly the same image that will run in production.  So you can still work with that image locally.

Multi stage building allows you to optionally target different stages of building that image when needed.

Reply:
I might just be overthinking it and misreading the examples you've provided!

I suppose I want to be sure that what CI builds and tests is the same thing that goes into production. My (quite possibly incorrect) reading of these is that the image used for testing is built from a different base than the production image.  That might not be a problem if all you're doing in this case is running your unit tests on a local development machine, but that's not explicit here and I would have expected tests to be run in CI. But then I can't really see why you'd need a different test image from development.

Reply:
That's probably my bad writing, they all run from the same base image.  If you look at the Node example each subsequent build stage is build on top of the previous until it gets the final image.  

Different tests in CI/CD may target different stages.  For example unit test will target the test stage but the majority of tests, security, accessibility, UI, performance etc will all be against the final stage when the file is run to completion.

With Node, you don't need a different base image for development.  

For .Net Core you (sometimes) do.  For .Net Core you should always run in production using an image that contains the .Net runtime which gives you just what you need to run in production.  But in order to actually compile a .Net Core application you need the .Net Core SDK which is several gigabytes in size.  For local development where you're developing and building in a container you'll need to use the SDK and not the runtime.  To create a runtime image you need an application to put in it, so you'll need to first build the application based on an SDK image then copy the built application to the runtime image.